src/asymmetric_crypto/ecies.rs

```diff
@@ -0,0 +1,313 @@
+///
+/// # Returns
+///
+/// * `Result<Vec<u8>, CryptoCoreError>` - The encrypted message as a `Vec<u8>`,
+///   or a `CryptoCoreError` if an error occurs.
+///
+/// # Example
+///
+/// ```
+/// use cosmian_crypto_core::{
+///    asymmetric_crypto::{curve25519::X25519KeyPair, DhKeyPair,ecies::ecies_encrypt},
+///    reexport::rand_core::SeedableRng,
+///    CsRng,
+/// };
+///
+/// let mut rng = CsRng::from_entropy();
+/// let key_pair: X25519KeyPair = X25519KeyPair::new(&mut rng);
+/// let msg = b"Hello, World!";
+///
+/// let _encrypted_message = ecies_encrypt::<
+///         CsRng,
+///         X25519KeyPair,
+///         { X25519KeyPair::PUBLIC_KEY_LENGTH },
+///         { X25519KeyPair::PRIVATE_KEY_LENGTH },
+///     >(
+///     &mut rng,
+///     &key_pair.public_key(),
+///     msg,
+///     None,
+///     None
+/// ).unwrap();
+/// ```
+#[allow(dead_code)]
+pub fn ecies_encrypt<R, DH, const PUBLIC_KEY_LENGTH: usize, const PRIVATE_KEY_LENGTH: usize>(
+    rng: &mut R,
+    receiver_public_key: &DH::PublicKey,
+    msg: &[u8],
+    shared_encapsulation_data: Option<&[u8]>,
+    shared_authentication_data: Option<&[u8]>,
+) -> Result<Vec<u8>, CryptoCoreError>
+where
+    R: CryptoRngCore,
+    DH: DhKeyPair<PUBLIC_KEY_LENGTH, PRIVATE_KEY_LENGTH>,
+    DH::PublicKey: From<DH::PrivateKey>,
+    for<'a, 'b> &'a DH::PublicKey: Add<&'b DH::PublicKey, Output = DH::PublicKey>
+        + Mul<&'b DH::PrivateKey, Output = DH::PublicKey>,
+    for<'a, 'b> &'a DH::PrivateKey: Add<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Sub<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Mul<&'b DH::PrivateKey, Output = DH::PrivateKey>
+        + Div<&'b DH::PrivateKey, Output = DH::PrivateKey>,
+{
+    // Generate an ephemeral key pair (r, R) where R = r.G
+    let ephemeral_key_pair = DH::new(rng);
+
+    // Calculate the shared secret point (Px, Py) = P = r.Y
+    let shared: DH::PublicKey = receiver_public_key * ephemeral_key_pair.private_key();
+
+    // S. Note: ECIES formally uses S = Px rather than the serialization of P
+    let mut shared_bytes = shared.to_bytes().to_vec();
+    // S||S1: if the user provided shared_encapsulation_data S1, then we append it
+    // to the shared_bytes S
+    if let Some(s1) = shared_encapsulation_data {
+        shared_bytes.extend(s1);
+    }
+
+    // Generate the 256-bit symmetric encryption key k, derived using Shake128 XOF
+    const SYMMETRIC_KEY_LENGTH: usize = Aes256GcmCrypto::KEY_LENGTH;
+    let key = kdf!(SYMMETRIC_KEY_LENGTH, &shared_bytes);
+
+    // Encrypt the message using AES-256-GCM
+    let nonce = <Aes256GcmCrypto as Dem<SYMMETRIC_KEY_LENGTH>>::Nonce::new(rng);
+
+    // Encrypt and authenticate the message, returning the ciphertext and MAC
+    let c_d = encrypt_combined(&key, msg, nonce.as_bytes(), shared_authentication_data)?;
+
+    // Assemble the final encrypted message: R || nonce || c || d
+    let mut ciphertext = ephemeral_key_pair.public_key().to_bytes().to_vec();
```

Comment:
Reserve the correct amount of space to avoid reallocations:
```
    let mut res =
        Vec::with_capacity(PUBLIC_KEY_LENGTH + NONCE_LENGTH + MAC_LENGTH + ciphertext.len());
    res.extend(ephemeral_key_pair.public_key().to_bytes().to_vec());
    res.extend(nonce.as_bytes());
    res.extend(&ciphertext);
    res.extend(tag);

    Ok(res)
```

Reply:
done